### DISCUSSION TITLE
No clear way to securely pass OIDC JWKS key in kubernetes secrets

### DISCUSSION THREAD

**Discussion by dlford** [authelia/authelia#6864] [Documentation Feedback] [answered]
This is a continuation of #6861.

If there is a way to securely pass OIDC JWKS keys in kubernetes secrets, it is not documented or not easy to find.

This does not work because `jwks` is an array:

https://www.authelia.com/configuration/methods/secrets/

> For instance the LDAP password can be defined in the configuration at the path authentication_backend.ldap.password, so this password could alternatively be set using the environment variable called AUTHELIA_AUTHENTICATION_BACKEND_LDAP_PASSWORD_FILE.

It is not a good idea to put secret values in the config map, so I don't see how this should be done properly.

For now `AUTHELIA_IDENTITY_PROVIDERS_OIDC_ISSUER_PRIVATE_KEY_FILE` still works, but that is deprecated.

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

Based on the information provided and the context from the sources, it seems you're encountering an issue with securely passing OIDC JWKS keys using Kubernetes secrets due to the `jwks` configuration being an array and the current documentation not clearly addressing this scenario. The use of `AUTHELIA_IDENTITY_PROVIDERS_OIDC_ISSUER_PRIVATE_KEY_FILE` is indeed deprecated as per the warning messages in the Authelia version 4.38.0, and the recommended approach is to use `jwks`.

For securely managing secrets in Kubernetes, including JWKS keys for Authelia, you can follow these general steps:

1. **Store the JWKS keys as Kubernetes secrets**: You can create a Kubernetes secret that contains your JWKS keys. Given that `jwks` is an array, you would typically store the entire JSON structure of the JWKS in a single secret key-value pair.

2. **Mount the secret into your Authelia Pod**: Use a volume mount to make the secret available inside your Authelia container. This is done by modifying your deployment configuration to include the secret volume and mount it to a specific path inside the container.

3. **Reference the secret file in Authelia's configuration**: Since you cannot directly use an environment variable to pass the JWKS array, you should reference the mounted secret file in your Authelia configuration file. This means you will have a `configuration.yml` file that includes a reference to the path where the secret is mounted, something like:
   ```yaml
   identity_providers:
     oidc:
       jwks: /path/to/mounted/secret/jwks.json
   ```
   Note: This step assumes that Authelia can load `jwks` from a file path, which might not be directly supported based on the current documentation. If Authelia does not support loading `jwks` from a file path, this would be a feature request or enhancement needed in Authelia.

Unfortunately, based on the provided context and sources, there is no direct documentation or example in Authelia's official documentation that outlines this process specifically for `jwks`. The closest related documentation involves using environment variables for secrets, but as you've noted, this method does not apply to array structures like `jwks`.

Given this, the best course of action would be to:
- Check Authelia's GitHub issues and discussions for any updates or workarounds provided by the community or the developers.
- Consider opening a new issue or discussion on the Authelia GitHub repository to seek clarification or request this feature if it's not currently supported.

This approach ensures that your secret values, including JWKS keys, are not exposed in config maps and are managed securely using Kubernetes secrets.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment, marked as answer]
See the example from the blog post. https://www.authelia.com/blog/4.38-release-notes/#changes-multiple-json-web-keys

**Reply by dlford** [reply]
I see, so there is no secure way to do it without using experimental features that "may break, be removed, or not work as expected"?

I'll stick with the old variable for now until this can be sorted out properly, thanks!

**Reply by james-d-elliott** [reply]
Use the old key, it still works. If you're going to be rude though you're not going to be welcome in the community.

**Reply by dlford** [reply]
I'm sorry, was not intending to be rude. You've been pretty short yourself.

I'm not sure what you mean by "use the old key"? I have a key... Maybe your're confused about the question?

**Reply by james-d-elliott** [reply]
> For now AUTHELIA_IDENTITY_PROVIDERS_OIDC_ISSUER_PRIVATE_KEY_FILE still works, but that is deprecated.

This. You already know it still works and that means you have read the message and it explains what the deprecation means, i.e. removal in 5.0.0. As far as documenting it we're not going to support it going forward and the template functionality is very likely to be the way we get around it.

Chances are like OpenID Connect 1.0 the breakage in the `template` function will be minimal if any (we've only had a single breaking change in the OpenID Connect 1.0 experimental phase being the token endpoint authentication method), the only time it's likely to change is in attempts to fix existing template functions or in adding new ones (or the potential removal of the `expand-env` filter).
